In [14]:
import pandas as pd
import numpy as np

In [15]:
%pip install pandas numpy

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.0.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [16]:
import pandas as pd
import numpy as np

In [17]:
print("Pandas version:", 
      pd.__version__)
print("Numpy version:",
      np.__version__)

Pandas version: 3.0.5
Numpy version: 2.5.2


In [18]:
df= pd.read_csv("student-por.csv", sep=";")

In [19]:
df = pd.read_csv("student-por.csv")

In [20]:
df.head()

,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,...,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
0,GP,F,18,U,GT3,A,4,4,at_home,teacher,...,4,3,4,1,1,3,4,0,11,11
1,GP,F,17,U,GT3,T,1,1,at_home,other,...,5,3,3,1,1,3,2,9,11,11
2,GP,F,15,U,LE3,T,1,1,at_home,other,...,4,3,2,2,3,3,6,12,13,12
3,GP,F,15,U,GT3,T,4,2,health,services,...,3,2,2,1,1,5,0,14,14,14
4,GP,F,16,U,GT3,T,3,3,other,other,...,4,3,2,1,2,5,0,11,13,13


In [21]:
df.shape

(649, 33)

In [22]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   school      649 non-null    str  
 1   sex         649 non-null    str  
 2   age         649 non-null    int64
 3   address     649 non-null    str  
 4   famsize     649 non-null    str  
 5   Pstatus     649 non-null    str  
 6   Medu        649 non-null    int64
 7   Fedu        649 non-null    int64
 8   Mjob        649 non-null    str  
 9   Fjob        649 non-null    str  
 10  reason      649 non-null    str  
 11  guardian    649 non-null    str  
 12  traveltime  649 non-null    int64
 13  studytime   649 non-null    int64
 14  failures    649 non-null    int64
 15  schoolsup   649 non-null    str  
 16  famsup      649 non-null    str  
 17  paid        649 non-null    str  
 18  activities  649 non-null    str  
 19  nursery     649 non-null    str  
 20  higher      649 non-null    str  
 21  inte

In [23]:
df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [24]:
df.duplicated().sum()

np.int64(0)

In [26]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   school      649 non-null    str  
 1   sex         649 non-null    str  
 2   age         649 non-null    int64
 3   address     649 non-null    str  
 4   famsize     649 non-null    str  
 5   Pstatus     649 non-null    str  
 6   Medu        649 non-null    int64
 7   Fedu        649 non-null    int64
 8   Mjob        649 non-null    str  
 9   Fjob        649 non-null    str  
 10  reason      649 non-null    str  
 11  guardian    649 non-null    str  
 12  traveltime  649 non-null    int64
 13  studytime   649 non-null    int64
 14  failures    649 non-null    int64
 15  schoolsup   649 non-null    str  
 16  famsup      649 non-null    str  
 17  paid        649 non-null    str  
 18  activities  649 non-null    str  
 19  nursery     649 non-null    str  
 20  higher      649 non-null    str  
 21  inte

In [27]:
df.describe()

,age,Medu,Fedu,traveltime,studytime,failures,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
count,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000,649.000000
mean,16.744222,2.514638,2.306626,1.568567,1.930663,0.221880,3.930663,3.180277,3.184900,1.502311,2.280431,3.536210,3.659476,11.399076,11.570108,11.906009
std,1.218138,1.134552,1.099931,0.748660,0.829510,0.593235,0.955717,1.051093,1.175766,0.924834,1.284380,1.446259,4.640759,2.745265,2.913639,3.230656
min,15.000000,0.000000,0.000000,1.000000,1.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,16.000000,2.000000,1.000000,1.000000,1.000000,0.000000,4.000000,3.000000,2.000000,1.000000,1.000000,2.000000,0.000000,10.000000,10.000000,10.000000
50%,17.000000,2.000000,2.000000,1.000000,2.000000,0.000000,4.000000,3.000000,3.000000,1.000000,2.000000,4.000000,2.000000,11.000000,11.000000,12.000000
75%,18.000000,4.000000,3.000000,2.000000,2.000000,0.000000,5.000000,4.000000,4.000000,2.000000,3.000000,5.000000,6.000000,13.000000,13.000000,14.000000
max,22.000000,4.000000,4.000000,4.000000,4.000000,3.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,32.000000,19.000000,19.000000,19.000000


In [29]:
for column in df.select_dtypes(include="object"
                ).columns:
    print(f"\n{column}:")
    print(df[column].unique())


school:
<StringArray>
['GP', 'MS']
Length: 2, dtype: str

sex:
<StringArray>
['F', 'M']
Length: 2, dtype: str

address:
<StringArray>
['U', 'R']
Length: 2, dtype: str

famsize:
<StringArray>
['GT3', 'LE3']
Length: 2, dtype: str

Pstatus:
<StringArray>
['A', 'T']
Length: 2, dtype: str

Mjob:
<StringArray>
['at_home', 'health', 'other', 'services', 'teacher']
Length: 5, dtype: str

Fjob:
<StringArray>
['teacher', 'other', 'services', 'health', 'at_home']
Length: 5, dtype: str

reason:
<StringArray>
['course', 'other', 'home', 'reputation']
Length: 4, dtype: str

guardian:
<StringArray>
['mother', 'father', 'other']
Length: 3, dtype: str

schoolsup:
<StringArray>
['yes', 'no']
Length: 2, dtype: str

famsup:
<StringArray>
['no', 'yes']
Length: 2, dtype: str

paid:
<StringArray>
['no', 'yes']
Length: 2, dtype: str

activities:
<StringArray>
['no', 'yes']
Length: 2, dtype: str

nursery:
<StringArray>
['yes', 'no']
Length: 2, dtype: str

higher:
<StringArray>
['yes', 'no']
Length: 2, dtype: 

C:\Users\josep\AppData\Local\Temp\ipykernel_19912\1962209225.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for column in df.select_dtypes(include="object"


In [30]:
df.select_dtypes(include="object"
                ).nunique()

C:\Users\josep\AppData\Local\Temp\ipykernel_19912\1356267474.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.select_dtypes(include="object"


school        2
sex           2
address       2
famsize       2
Pstatus       2
Mjob          5
Fjob          5
reason        4
guardian      3
schoolsup     2
famsup        2
paid          2
activities    2
nursery       2
higher        2
internet      2
romantic      2
dtype: int64

In [31]:
cleaned_df = df.copy()

In [35]:
cleaned_df.columns = (
    cleaned_df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

In [39]:
cleaned_df.columns.tolist()

['school',
 'sex',
 'age',
 'address',
 'famsize',
 'pstatus',
 'medu',
 'fedu',
 'mjob',
 'fjob',
 'reason',
 'guardian',
 'traveltime',
 'studytime',
 'failures',
 'schoolsup',
 'famsup',
 'paid',
 'activities',
 'nursery',
 'higher',
 'internet',
 'romantic',
 'famrel',
 'freetime',
 'goout',
 'dalc',
 'walc',
 'health',
 'absences',
 'g1',
 'g2',
 'g3']

In [45]:
for column in cleaned_df.select_dtypes(include="object").columns:
    cleaned_df[column] = cleaned_df[column].str.strip()

C:\Users\josep\AppData\Local\Temp\ipykernel_19912\1365292956.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for column in cleaned_df.select_dtypes(include="object").columns:


In [46]:
cleaned_df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
pstatus       0
medu          0
fedu          0
mjob          0
fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
dalc          0
walc          0
health        0
absences      0
g1            0
g2            0
g3            0
dtype: int64

In [47]:
cleaned_df.duplicated().sum()

np.int64(0)

In [49]:
cleaned_df.dtypes

school          str
sex             str
age           int64
address         str
famsize         str
pstatus         str
medu          int64
fedu          int64
mjob            str
fjob            str
reason          str
guardian        str
traveltime    int64
studytime     int64
failures      int64
schoolsup       str
famsup          str
paid            str
activities      str
nursery         str
higher          str
internet        str
romantic        str
famrel        int64
freetime      int64
goout         int64
dalc          int64
walc          int64
health        int64
absences      int64
g1            int64
g2            int64
g3            int64
dtype: object

In [51]:
print("Age:",cleaned_df["age"].min(), "to", cleaned_df["age"].max())
print("G1:",cleaned_df["g1"].min(),"to", cleaned_df["g1"].max())
print("G2:",cleaned_df["g2"].min(),"to", cleaned_df["g2"].max())
print("G3:",cleaned_df["g3"].min(),"to", cleaned_df["g3"].max())
print("Failure:",cleaned_df["failures"].min(), "to", cleaned_df["failures"].max())

Age: 15 to 22
G1: 0 to 19
G2: 0 to 19
G3: 0 to 19
Failure: 0 to 3


In [53]:
cleaned_df.shape

(649, 33)

In [54]:
cleaned_df.head()

,school,sex,age,address,famsize,pstatus,medu,fedu,mjob,fjob,...,famrel,freetime,goout,dalc,walc,health,absences,g1,g2,g3
0,GP,F,18,U,GT3,A,4,4,at_home,teacher,...,4,3,4,1,1,3,4,0,11,11
1,GP,F,17,U,GT3,T,1,1,at_home,other,...,5,3,3,1,1,3,2,9,11,11
2,GP,F,15,U,LE3,T,1,1,at_home,other,...,4,3,2,2,3,3,6,12,13,12
3,GP,F,15,U,GT3,T,4,2,health,services,...,3,2,2,1,1,5,0,14,14,14
4,GP,F,16,U,GT3,T,3,3,other,other,...,4,3,2,1,2,5,0,11,13,13


In [55]:
print("Original shape:",df.shape)
print("Cleaned shape:",cleaned_df.shape)

Original shape: (649, 33)
Cleaned shape: (649, 33)


In [56]:
cleaned_df.to_csv("cleaned_student_por.csv", index=False)

In [57]:
import os
print(os.path.exists("cleaned_student_por.csv"))

True


In [58]:
print("FINAL DATA QUALITY CHECK")
print("-------------------------")
print("Rows and columns:",cleaned_df.shape)
print("Missing values:",cleaned_df.isnull().sum().sum())
print("Duplicate rows:",cleaned_df.duplicated().sum())
print("columns:",len(cleaned_df.columns))

FINAL DATA QUALITY CHECK
-------------------------
Rows and columns: (649, 33)
Missing values: 0
Duplicate rows: 0
columns: 33


# Student Performance Data Cleaning

## Dataset
The dataset contains student information and academic performance records.

## Data Cleaning Steps
The following steps were performed:

1. Loaded the original CSV dataset using Pandas.
2. Inspected the dataset structure, columns, data types and statistical summary.
3. Checked for missing values.
4. Checked for duplicate records.
5. Checked categorical and numerical values for consistency.
6. Standardized column names by removing unnecessary spaces, converting them to lowercase and replacing spaces with underscores.
7. Removed unnecessary leading and trailing spaces from text values.
8. Validated numerical values such as age, grades and number of failures.
9. Verified that the cleaned dataset retained 649 rows and 33 columns.
10. Saved the cleaned dataset as `cleaned_student_por.csv`.

## Data Quality Result
The dataset contained no missing values and no duplicate rows. The numerical values were within reasonable ranges, so no valid records were unnecessarily removed.

The cleaned dataset was saved as `cleaned_student_por.csv`.

In [62]:
cleaned_df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
pstatus       0
medu          0
fedu          0
mjob          0
fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
dalc          0
walc          0
health        0
absences      0
g1            0
g2            0
g3            0
dtype: int64